File: java/Lists/Stack.java
package Lists;

public class Stack<E> {
    private DoubleLinkedList<E> list;
    private Integer maxCapacity;

    public Stack() {
        this.list = new DoubleLinkedList<>();
        maxCapacity = null;
    }

    public Stack(int maxCapacity) {
        this.list = new DoubleLinkedList<>();
        this.maxCapacity = maxCapacity;
    }

    public void push(E val) {
        if (maxCapacity != null && list.size() == maxCapacity) throw new RuntimeException("Stack overflow");
        list.addLast(val);
    }

    public int size() {
        return list.size();
    }

    public E peek() {
        return list.getLast();
    }

    public E pop() {
        return list.removeLast();
    }

    public void clear() {
        list.clear();
    }

    public boolean isEmpty() {
        return list.isEmpty();
    }
}

File: typescript/src/lists/stack.ts
import DoubleLinkedList from "./double-linked-list.ts";

export default class Stack<E> {
  private list: DoubleLinkedList<E>;
  private maxCapacity: number;

  constructor(maxCapacity: number = -1) {
    this.list = new DoubleLinkedList();
    this.maxCapacity = maxCapacity;
  }

  public push(val: E) {
    if (this.maxCapacity != -1 && this.list.size == this.maxCapacity) {
      throw new Error("Stack overflow");
    }
    this.list.addLast(val);
  }

  public get size() {
    return this.list.size;
  }

  public peek() {
    return this.list.getLast();
  }

  public pop() {
    return this.list.removeLast();
  }

  public clear() {
    this.list.clear();
  }
}
